### PR TITLE
Refactor: TokenServer with DerivedState

### PR DESCRIPTION
**Pull request by y-myajima**
# Description

_description_with_details_and_reasoning_

## Assignees
- [x] `Assignees` is set

## Type of changes
- [x] Apply one or more `labels` of the following that fits:
  - `bug`: Bug fix
  - `dependencies`: Dependency upgrades
  - `documentation`: Documentation changes
  - `enhancement`: New Feature
  - `good first issue`: First contribution
  - `logging`: Log changes
  - `refactor`: Refactoring (no functional changes, no api changes)

### Flags
```
- [ ] Breaks backward compatibility
- [ ] Requires a documentation update
- [ ] Has untestable code
```
---

## Checklist
```
- [ ] Followed the guidelines in the CONTRIBUTING document
- [ ] Added prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]` in the PR title if necessary
- [ ] Tested and linted the code
- [ ] Commented the code
- [ ] Made corresponding changes to the documentation
```

## Checklist for maintainer
```
- [ ] Use `Squash and merge`
- [ ] Double-confirm the merge message has prefix `[skip ci]`/`[ci skip]`/`[no ci]`/`[skip actions]`/`[actions skip]`
- [ ] Delete the branch after merge
```

### CODE REVIEW

**Review by mlajkim — COMMENTED** [review]
review

File: pkg/config/derived-token-server.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+type HeaderTokenMode struct {
+	Use            bool
+	RoleAuthHeader string
+}
+
+type RestAPIMode struct {
+	Use bool
+}
```
[the remark this answers is not in the document]
Comment (mlajkim) [reply pkg/config/derived-token-server.go:30]: nah.

File: pkg/config/derived-token-server.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+type DerivedTokenServer struct {
+	Use             bool
+	Addr            string
+	ShutdownDelay   time.Duration
+	ShutdownTimeout time.Duration
+	ServerTimeout   time.Duration
+	TLS             TLS
+	HeaderToken     HeaderTokenMode
+	RestAPI         RestAPIMode
```
[the remark this answers is not in the document]
Comment (mlajkim) [reply pkg/config/derived-token-server.go:47]: mode on the top right before the `Use`

File: pkg/config/derived-token-server.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+			Use: false,
+		},
+	}
+
+	if idCfg.Init {
+		return nil // disabled
+	}
+	if idCfg.tokenServerAddr == "" || (!strings.Contains(idCfg.TokenType, "accesstoken") && !strings.Contains(idCfg.TokenType, "roletoken")) {
+		return nil // disabled
+	}
```
[the remark this answers is not in the document]
Comment (mlajkim) [reply pkg/config/derived-token-server.go:79]: optional

File: pkg/config/derived-token-server.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (y-myajima) [reply pkg/config/derived-token-server.go]: handled in: https://github.com/AthenZ/k8s-athenz-sia/pull/161/commits/f2e5789407d30f168ca66ca2d011d28392e80167

**Review by mlajkim — APPROVED** [review]
ty

**Pull request merged**